src/tdigest.rs

```diff
@@ -0,0 +1,269 @@
+    }
+
+    // shamefully stolen from the better tdigest rust crate
+    /// Estimates the value at the given quantile.
+    pub fn estimate_value(&self, q: f32) -> f32 {
+        if self.centroids.len() == 0 {
+            return 0.0;
+        }
+
+        let count_ = self.centroids.iter().fold(0, |sum, x| sum + x.weight) as f32;
+        let rank = count_ * q;
+
+        let self_min = self.centroids[0].mean;
+        let self_max = self.centroids.last().expect("uh oh").mean;
+
+        // the following chunk of code does the same thing as search_centroids but it's based on quantile instead of mean and also it's much faster for the use case (most queries are going to be at the extremes, so linear search works out to be pretty fast)
+        let mut pos: usize;
+        let mut t: f32;
+        if q > 0.5 {
+            if q >= 1.0 {
+                return self_max;
+            }
+
+            pos = 0;
+            t = count_;
+
+            for (k, centroid) in self.centroids.iter().enumerate().rev() {
+                t -= centroid.weight as f32;
+
+                if rank >= t {
+                    pos = k;
+                    break;
+                }
+            }
+        } else {
+            if q <= 0.0 {
+                return self_min;
+            }
+
+            pos = self.centroids.len() - 1;
+            t = 0.0;
+
+            for (k, centroid) in self.centroids.iter().enumerate() {
+                if rank < t + centroid.weight as f32 {
+                    pos = k;
+                    break;
+                }
+
+                t += centroid.weight as f32;
+            }
+        }
+
+        // TODO: document this lmao
+        let mut delta = 0.0;
+        let mut min: f32 = self_min.clone();
+        let mut max: f32 = self_max.clone();
+
+        if self.centroids.len() > 1 {
+            if pos == 0 {
+                delta = self.centroids[pos + 1].mean - self.centroids[pos].mean;
+                max = self.centroids[pos + 1].mean;
+            } else if pos == (self.centroids.len() - 1) {
+                delta = self.centroids[pos].mean - self.centroids[pos - 1].mean;
+                min = self.centroids[pos - 1].mean;
+            } else {
+                delta = (self.centroids[pos + 1].mean - self.centroids[pos - 1].mean) / 2.0;
+                min = self.centroids[pos - 1].mean;
+                max = self.centroids[pos + 1].mean;
+            }
+        }
+
+        let value = self.centroids[pos].mean + ((rank - t) / self.centroids[pos].weight as f32 - 0.5) * delta;
+        Self::clamp(value, min, max)
+    }
+    
+    // wanted to add a sorted version of this but i couldn't think of a version to save that much time and this is pretty fast already. if you wanted to make it super duper fast that's what you would do i guess
+    /// Merges an unsorted vector of values into the T-Digest.
+    pub fn merge_vec(&mut self, v: Vec<f32>) {
+        self.centroids.extend(Self::vals_to_centroids(v));
```

Comment:
true, but this function is supposed to be public, and i feel that it would be more awkward to ask the user to pass in an `IntoIterator` than have to do a quick allocation before merging (especially since the idea behind the buffer is an occasional expensive draining while being fast most of the time).
i will see if i can figure something out for the `vals_to_centroids` thing though

Reply:
I agree with david, it's usually better to allow any kind of iterator if it doesn't cost anything, and in this case you also gain something.

From the public api point of view, users can still pass a `Vec` if they have it already, but if they don't you allow them to pass an iterator too.

This function can accept an `impl IntoIterator<Item = f32>`, so can `Self::vals_to_centroid` (you change from `.iter` to `.into_iter`), and then you can change the return type of `vals_to_centroid` to `impl IntoIterator<Item = Centroid>`, and avoid the `collect` since the `extend` here accepts an iterator too.